### ISSUE TITLE
wxPython fails to properly add the right modules, forcing adding it to 'packages' option

### ISSUE THREAD

**Issue by urishab** [py2exe/py2exe#7]
A new issue where py2exe fails to add the correct modules of wxPython.
Not sure if this is the result of the new py2exe or the new wxPython but in previous versions it did not occur.

My test7.py:
```
from wx.richtext import RichTextCtrl
print('hello')

```
My py2exe_test7.py:
```
from distutils.core import setup
import py2exe #@UnusedImport
setup( console=[{ "script": "test7.py",},],
       options={"py2exe": {
              "includes":['imp'],
       }})

```
If trying to run test7.exe I get the folowing error:
```
Traceback (most recent call last):
  File "test7.py", line 1, in <module>
  File "C:\Projects\py2exe\pyenv\lib\site-packages\wx\richtext.py", line 17, in <module>
    from ._richtext import *
  File "<loader>", line 10, in <module>
  File "<loader>", line 8, in __load
ImportError: (No module named 'wx._adv') 'C:\\Projects\\py2exe\\dist\\wx._richtext.pyd'

```
To solve this I add wx to packages like this:
```
from distutils.core import setup
import py2exe #@UnusedImport
setup( console=[{ "script": "test7.py",},],
       options={"py2exe": {
              "includes":['imp'],
              'packages':['wx',]
       }})
```

It solves the issue but results in many unnecessary files in the dist folder

My pip freeze:
```
numpy==1.16.2
Pillow==6.0.0
py2exe==0.9.3.1
pypiwin32==223
pywin32==224
six==1.12.0
wxPython==4.0.4
```

**Comment by albertosottile** [comment]
> Not sure if this is the result of the new py2exe or the new wxPython but in previous versions it did not occur.

As far as I know, `py2exe` never correctly imported the dependent submodules of graphics packages (or large packages in general), e.g. wx, PySide, PyQt. To solve this issue, the solution is to import the whole package, as you did, and manually remove the files that you think are not needed from the `dist` folder. This can be done manually or in the `setup.py` script, as done for example [here](https://github.com/Syncplay/syncplay/blob/5a9e2cde23804e623e311d87398861cd1862c583/buildPy2exe.py#L616). 

As much as I would like to have an automatic subpackage-aware module finder for these packages, like the one used by PyInstaller, I do not have the time to implement this at the moment. Of course, PRs are always welcome.

**Comment by urishab** [comment]
OK fair enough
Regarding the module finder, I thought that was what the hooks are for. Aren't they submodule specific?

**Comment by albertosottile** [comment]
Sure, a hook could be a way to implement this. However, a detailed list of the internal dependencies for each submodule would be required to write a hook that solves this and similar issues. 

It would be much better to upgrade the module finder to allow finding these dependencies automatically (e.g. understanding that importing `wx.richtext` requires to import `wx.adv`), but that takes time.

**Comment by leonardder** [comment]
In [NVDA](https://github.com/nvaccess/nvda/), we're using py2exe 0.6.9 with wxPython, and we don't have to list wx in the packages list. In other words, that version of py2exe seems to be able to find out what parts of the wx package are required.
Our setup script imports wx at the top though, but I don't think that will make a difference.

**Comment by albertosottile** [comment]
There's something that py2exe is able to identify and something that is not. This is both due to how modulefinder works and due to how some internal imports happen in wx. I guess an alternative could be to map just the modules that are not imported correctly and write a hook accordingly, but that still takes time.

@leonardder It looks like your setup script imports the whole wx package. In that case, there will be no need for a hook or a module walker. What @urishab was trying to do was just to import `wx.richtext` and make it work. This is not possible unless py2exe knows somehow that also `wx.adv` is required, either automatically (modulefinder) or manually (a hook). Of course, a workaround to this issue could be to just import the whole package, as you do, and then delete the unwanted modules from the packed executable (if needed). This is what I have done so far with `PySide2` (Qt for Python) in my projects.

@leonardder Did you try py2exe 0.9.3.1 with NVDA?

**Comment by leonardder** [comment]
Thanks for your quick reply, I think we'll just have to test this out.

> @leonardder Did you try py2exe 0.9.3.0 with NVDA?

I'm currently rewriting the initial description of nvaccess/nvda#8375, and then I'll have a discussion with NV Access about what path to follow.

**Comment by urishab** [comment]
I did a little digging into the code that analyzes the wxPython package to look for dependencies.
It seems the wxPython is built as a c code  PYD package. That means that it cannot be analyzed for "import" statements using the python disassembler because they are in c.

However the wxPython contains "stub" files (*.pyi) which are sort of interface files for python. 
For example the _richtext.pyi file just contains 3 import statements which are the exact dependencies of _richtext.PYD and also the headers of all the functions exported by _richtext.PYD.

So if one were to analyze a package like wxPython correctly they would simply need, for every PYD file analyzed, load the corresponding PYI file and analyze it instead.
Seems to be not too hard.

I'm not sure how prevalent and standard this practice is. It seems to be mainly for the purpose of correct introspection for packages such as py2exe.
The only information I could find about these files is in [PEP484 ](https://www.python.org/dev/peps/pep-0484/)

**Comment by albertosottile** [comment]
This seems great, thank you for researching it. I will definitely give these files and PEP 484 a look when I have the time to work on this enhancement. In the meantime, if you want to give it a try, PRs are always welcome.

**Comment by albertosottile** [comment]
**TL,DR**: I am closing this issue as stub files are not meant to be used for this purpose and, indeed, they would not work for e.g. PySide2. Solution to this bug is to import the whole package in the bundle and, if desired, to manually delete unneeded libraries. Nevertheless, the new `dllfinder` introduced in v0.10.x mitigates but does not solve this issue completely.

-------

After some digging, I determined that the stub-only files defined by PEP 484 and PEP 561 are not suitable for the purpose of identifying imports that occurs outside of Python (e.g. in C code). 

First of all, tracing imports or drawing an interface to Python is not the purpose of these files. According to the two PEPs, the purpose of these files is to store the type hints information of the related module:
> Stub files are files containing type hints that are only for use by the type checker

So, while it is reasonable to assume that all the imports found in the stub files are indeed imports that occur in the relative code, there is no guarantee that these are the **only** imports performed by the code. There might be further internal imports that do not affect the interface exposed to Python nor the Python typing which will not included in the stub files.

Indeed, before starting any implementation, I looked at the use I make of PySide2 in another software to check if the stub files would have helped me in finding the right dependencies. In this application, we explicitly import only `PySide2.QtCore`, `PySide2.QtGui`, and `PySide2.QtWidgets`. Looking at the three stub files of these modules, these are the relevant import statements:

- QtCore.pyi
```python
import PySide2.QtCore
``` 

- QtGui.pyi
```python
import PySide2.QtCore
import PySide2.QtGui
``` 

- QtWidgets.pyi
```python
import PySide2.QtCore
import PySide2.QtGui
import PySide2.QtWidgets
``` 

So, by looking at the stub files one might be tempted to say that this set of modules should be self consistent: they only import themselves so, including just these three in the packed executable should be enough to distribute it safely. 

However, reality is different. To let the executable run, at least these libraries are needed (this was determined by including all the libraries and then deleting them one by one until the application could not start anymore):

- Windows
```Qt5Core.dll, Qt5Gui.dll, Qt5Network.dll, Qt5PositioningQuick.dll, Qt5Qml.dll, Qt5RemoteObjects.dll, Qt5Widgets.dll```

- macOS
```QtCore.framework, QtDBus.framework, QtGui.framework, QtNetwork.framework, QtPrintSupport.framework, QtQml.framework, QtWidgets.framework```

(This investigation also shows that the library set is actually platform-dependent, something that could never be reflected in the stub files.)

In summary, parsing stub files for hidden import, while it could be useful in some cases (like the one reported in this issue),  will not solve this problem in the general case. Hence, I will not invest more time in developing a modulefinder based on these files. 

So, I am closing this ticket. The short term solution for this issue is to adapt the `setup.py` file to include the missing dependency. Medium term solutions could be to write a hook that imports everything and, if needed, manually delete unneeded libraries from the packed bundle (following the approach that I use for PySide2 and also pursued in #39 for PyQt5). 

A long term approach would be to improve the new `dllfinder` based on `pefile` (introduced in version 0.10.x) which should be better in finding the correct linkage from the external code. I did run this test again with 0.10.1.0 and observed that the executable still fails because it cannot find `wx._xml`. This time, though, `py2exe` correctly identifies that `wx._richtest` requires `wx._adv, wx._html, wx._msw, wx._core, wx._siplib` and other stuff. Perhaps just adding `wx._xml` via a hook could help in this case, but this still does not provide a general solution. As why this link is not found by the new finder, I honestly have no idea...

**Comment by albertosottile** [comment]
> As why this link is not found by the new finder, I honestly have no idea...

Further investigation revealed that the hidden import happens in the `wx\_richtext.cp38-win_amd64.pyd` file hence, it is not visible to `ModuleFinder`. On top of that, the `DllFinder` does not identify a link to `wx._xml` because the library does not link to it directly but, instead, uses a quite complicated way to do so based on an internal system called `etgtools`. Therefore, it seems to me that a hook is the most sensible solution to this specific issue.

**Comment by urishab** [comment]
OK thanks for diving into this!
It sounds like a hook is the right solution.
